src/tr2/decomp/skidoo.c

```diff
@@ -0,0 +1,1039 @@
+            g_LaraItem->goal_anim_state = LARA_STATE_SKIDOO_LEFT;
+        } else if (g_Input & IN_RIGHT) {
+            g_LaraItem->goal_anim_state = LARA_STATE_SKIDOO_RIGHT;
+        }
+        break;
+
+    case LARA_STATE_SKIDOO_LEFT:
+        if (!(g_Input & IN_LEFT)) {
+            g_LaraItem->goal_anim_state = LARA_STATE_SKIDOO_SIT;
+        }
+        break;
+
+    case LARA_STATE_SKIDOO_RIGHT:
+        if (!(g_Input & IN_RIGHT)) {
+            g_LaraItem->goal_anim_state = LARA_STATE_SKIDOO_SIT;
+        }
+        break;
+
+    case LARA_STATE_SKIDOO_FALL:
+        if (skidoo->fall_speed <= 0 || skidoo_data->left_fallspeed <= 0
+            || skidoo_data->right_fallspeed <= 0) {
+            Sound_Effect(SFX_CLATTER_3, &skidoo->pos, SPM_NORMAL);
+            g_LaraItem->goal_anim_state = LARA_STATE_SKIDOO_SIT;
+        } else if (skidoo->fall_speed > DAMAGE_START + DAMAGE_LENGTH) {
+            g_LaraItem->goal_anim_state = LARA_STATE_SKIDOO_LET_GO;
+        }
+        break;
+
+    case LARA_STATE_SKIDOO_STILL: {
+        const int32_t music_track =
+            M_IsArmed(skidoo_data) ? MX_BATTLE_THEME : MX_SKIDOO_THEME;
+        if (!(g_MusicTrackFlags[music_track] & IF_ONE_SHOT)) {
+            Music_Play(music_track, false);
+            g_LaraItem = g_LaraItem;
+            g_MusicTrackFlags[music_track] |= IF_ONE_SHOT;
+        }
+
+        if (dead) {
+            g_LaraItem->goal_anim_state = LARA_STATE_SKIDOO_DEATH;
+            return;
+        }
+
+        g_LaraItem->goal_anim_state = LARA_STATE_SKIDOO_STILL;
+
+        if (g_Input & IN_JUMP) {
+            if ((g_Input & IN_RIGHT)
+                && Skidoo_CheckGetOffOK(LARA_STATE_SKIDOO_GET_OFF_R)) {
+                g_LaraItem->goal_anim_state = LARA_STATE_SKIDOO_GET_OFF_R;
+                skidoo->speed = 0;
+            } else if (
+                (g_Input & IN_LEFT)
+                && Skidoo_CheckGetOffOK(LARA_STATE_SKIDOO_GET_OFF_L)) {
+                g_LaraItem->goal_anim_state = LARA_STATE_SKIDOO_GET_OFF_L;
+                skidoo->speed = 0;
+            }
+        } else if (g_Input & IN_LEFT) {
+            g_LaraItem->goal_anim_state = LARA_STATE_SKIDOO_LEFT;
+        } else if (g_Input & IN_RIGHT) {
+            g_LaraItem->goal_anim_state = LARA_STATE_SKIDOO_RIGHT;
+        } else if (g_Input & (IN_BACK | IN_FORWARD)) {
+            g_LaraItem->goal_anim_state = LARA_STATE_SKIDOO_SIT;
+        }
+        break;
+    }
+
+    default:
+        break;
+    }
+}
+
+void __cdecl Skidoo_Explode(const ITEM *const skidoo)
+{
+    const int16_t fx_num = Effect_Create(skidoo->room_num);
+    if (fx_num != NO_ITEM) {
+        FX *const fx = &g_Effects[fx_num];
+        fx->pos.x = skidoo->pos.x;
+        fx->pos.y = skidoo->pos.y;
+        fx->pos.z = skidoo->pos.z;
+        fx->speed = 0;
+        fx->frame_num = 0;
+        fx->counter = 0;
+        fx->object_id = O_EXPLOSION;
+    }
+
+    Effect_ExplodingDeath(g_Lara.skidoo, ~SKIDOO_GUN_MESH, 0);
```

Comment:
If either the red or black skidoo explodes, you see the main part of the red mesh appear. In OG, you only see "side bits" of the model fly off. Possibly the `mesh_bits` here aren't quite right for this call?

![image](https://github.com/user-attachments/assets/1cb5ce94-e9eb-4dba-830d-b90f4a428c14)

Reply:
Thanks – the value indeed was wrong, rather than ~4 it should've been -4 or ~(4 - 1). I have pushed a fix, but I haven't tested it.

Reply:
BTW the name `SKIDOO_GUN_MESH` I think is wrong and should be changed – but I don't fully understand its usage. For sure it has to do with the track mesh, but I'm not certain as to why `track_mesh & 4` yields a truthy value for armed snowmobiles and a falsy value for normal snowmobiles.